libparsec/crates/platform_device_loader/src/web/internal.rs

```diff
@@ -1,103 +1,111 @@
 
-use std::{path::Path, sync::Arc};
+use std::{path::Path, rc::Rc, sync::Arc};
 
-use itertools::Itertools;
+use libparsec_platform_async::{
+    lock::Mutex,
+    stream::{StreamExt, TryStreamExt},
+};
 use libparsec_types::{
     AvailableDevice, DateTime, DeviceAccessStrategy, DeviceFile, DeviceFilePassword, LocalDevice,
 };
 
-use super::error::*;
+use crate::web::wrapper::{DirEntry, DirOrFileHandle, OpenOptions};
+
+use super::{
+    error::*,
+    wrapper::{Directory, File},
+};
 
 pub struct Storage {
-    storage: web_sys::Storage,
+    root_dir: Directory,
 }
 
 impl Storage {
-    pub(crate) fn new() -> Result<Self, NewStorageError> {
-        let window = web_sys::window().ok_or(NewStorageError::NoWindow)?;
-        let storage = if cfg!(test) {
-            // Use session storage for tests, So the browser does not try to persist the data.
-            window.session_storage()
-        } else {
-            window.local_storage()
-        }
-        .map_err(NewStorageError::WindowError)
-        .and_then(|v| v.ok_or(NewStorageError::NoLocalStorage))?;
-        Ok(Self { storage })
+    pub(crate) async fn new() -> Result<Self, NewStorageError> {
+        let root_dir = Directory::get_root_dir().await?;
+        Ok(Self { root_dir })
     }
 
-    pub(crate) fn list_available_devices(
+    pub(crate) async fn list_available_devices(
         &self,
         config_dir: &Path,
     ) -> Result<Vec<AvailableDevice>, ListAvailableDevicesError> {
         let devices_dir = crate::get_devices_dir(config_dir);
-        let device_prefix = format!("{}/", devices_dir.display());
-        log::debug!("Will list devices starting with `{device_prefix}`");
-        let items_count =
-            self.storage
-                .length()
-                .map_err(|e| ListAvailableDevicesError::GetItemStorage {
-                    key: "length".to_owned(),
-                    error: e,
-                })?;
-        (0..items_count)
-            .filter_map(|i| {
-                let key = self.storage.key(i).ok().flatten();
-                match key {
-                    Some(key)
-                        if key.starts_with(&device_prefix)
-                            && key.ends_with(crate::DEVICE_FILE_EXT) =>
+        log::debug!(
+            "Will list devices starting with `{}` and having suffix {}",
+            devices_dir.display(),
+            crate::DEVICE_FILE_EXT
+        );
+        let dir = match self
+            .root_dir
+            .get_dir_handle_from_path(&devices_dir, None)
+            .await
+        {
+            Ok(dir) => dir,
+            Err(GetDirectoryHandleError::NotFound { .. }) => {
+                log::debug!("Could not found devices dir");
+                return Ok(Vec::new());
+            }
+            Err(e) => return Err(e.into()),
+        };
+        let mut devices = Vec::<AvailableDevice>::new();
+        let dirs_to_explore = Rc::new(Mutex::new(Vec::from([dir])));
```

Comment:
Nitpick: import `libparsec_async::lock::Mutex` as `AsyncMutex` (this is what is done in the rest of the codebase and makes it obvious we are not dealing with the sync version)

Reply:
BTW, the `RC<Mutex<>>` is not needed here since the variable only lives in this function (hence it is never shared).

Reply:
done